include/pgduckdb/logger.hpp

```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+extern "C" {
+bool errstart(int elevel, const char *domain);
+void errfinish(const char *filename, int lineno, const char *funcname);
+int	errmsg_internal(const char *fmt,...);
+}
+
+namespace pgduckdb {
+
+/* PG Error level codes */
+#define DEBUG5		10
+#define DEBUG4		11
+#define DEBUG3		12
+#define DEBUG2		13
+#define DEBUG1		14
+#define LOG			15
+#define INFO		17
+#define NOTICE		18
+#define WARNING		19
+#define PGWARNING	19
+#define WARNING_CLIENT_ONLY	20
+
+// From PG elog.h
+#if defined(errno) && defined(__linux__)
+#define pd_prevent_errno_in_scope() int __errno_location pg_attribute_unused()
+#elif defined(errno) && (defined(__darwin__) || defined(__FreeBSD__))
+#define pd_prevent_errno_in_scope() int __error pg_attribute_unused()
+#else
+#define pd_prevent_errno_in_scope()
+#endif
+
+#define pd_ereport_domain(elevel, domain, ...)	\
+	do { \
+		pd_prevent_errno_in_scope(); \
+		static_assert(elevel >= DEBUG5 && elevel <= WARNING_CLIENT_ONLY, "Invalid error level"); \
+		if (errstart(elevel, domain)) \
+			__VA_ARGS__, errfinish(__FILE__, __LINE__, __func__); \
```

Comment:
One thing I'm now realizing: I think we cannot call `errstart` safely without the global lock. It internally calls `get_error_stack_entry` which increments a global counter... (which ofcourse is not an atomic increment). Also `errfinish` is using `appendStringInfo`, which does allocations, and the postgres allocator is not thread safe. So basically what I'm saying: We cannot safely log without a global lock. 

We could safely check the `elevel` though without the lock, using `message_level_is_interesting()`. And only take the lock when the that function returns true.

Reply:
That's a great point. I didn't pay attention to that yet. I'll rework it to make it thread-safe.